src/main/java/com/leff/midi/MidiFile.java

```diff
@@ -0,0 +1,106 @@
+//////////////////////////////////////////////////////////////////////////////
+//	Copyright 2011 Alex Leffelman
+//
+//	Licensed under the Apache License, Version 2.0 (the "License");
+//	you may not use this file except in compliance with the License.
+//	You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+//	Unless required by applicable law or agreed to in writing, software
+//	distributed under the License is distributed on an "AS IS" BASIS,
+//	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//	See the License for the specific language governing permissions and
+//	limitations under the License.
+//////////////////////////////////////////////////////////////////////////////
+
+package com.leff.midi;
+
+import com.leff.midi.util.MidiUtil;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class MidiFile {
+    public static final int HEADER_SIZE = 14;
+    public static final byte[] IDENTIFIER = {'M', 'T', 'h', 'd'};
+
+    public static final int DEFAULT_RESOLUTION = 480;
+
+    private int mType;
+    private int mTrackCount;
+    private int mResolution;
+    private List<MidiTrack> mTracks = new ArrayList<>();
+
+    public MidiFile(byte[] buffer) throws IOException {
+        this(new ByteArrayInputStream(buffer));
+    }
+
+    public MidiFile(InputStream in) throws IOException {
+        readHeader(in);
+        for (int i = 0; i < mTrackCount; i++) {
+            mTracks.add(new MidiTrack(in));
+        }
+    }
+
+    private void readHeader(InputStream in) throws IOException {
+        byte[] buffer = new byte[HEADER_SIZE];

```

Comment:
Many candidates seemed to get confused with the proliferation of objects called `buffer`, especially since in the previous version this `buffer` continued to be in scope (needlessly) even while parsing the track objects.  Pushing this into its own function ensures that the scope of this `buffer` instance is limited to while we're parsing the header.